tpm/tpm.go

```diff
@@ -0,0 +1,185 @@
+	}
+	glog.Infof("PCR %v value: 0x%x\n", pcr, pcrVal)
+
+	// Get the authorization policy that will protect the data to be sealed
+	sessHandle, policy, err := policyPCRPasswordSession(rwc, pcr, objectPassword)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to get policy: %v", err)
+	}
+	if err := tpm2.FlushContext(rwc, sessHandle); err != nil {
+		return nil, nil, fmt.Errorf("unable to flush session: %v", err)
+	}
+	glog.Infof("Created authorization policy: 0x%x\n", policy)
+
+	// Seal the data to the parent key and the policy
+	privateArea, publicArea, err := tpm2.Seal(rwc, srkHandle, srkPassword, objectPassword, policy, dataToSeal)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to seal data: %v", err)
+	}
+	glog.Infof("Sealed data: 0x%x\n", privateArea)
+
+	return privateArea, publicArea, nil
+}
+
+// Returns the unsealed data
+func Unseal(tpmPath string, pcr int, srkPassword, objectPassword string, privateArea, publicArea []byte) ([]byte, error) {
+	// Open the TPM
+	rwc, err := tpm2.OpenTPM(tpmPath)
+	if err != nil {
+		return nil, fmt.Errorf("can't open TPM %q: %v", tpmPath, err)
+	}
+	defer func() {
+		rwc.Close()
+	}()
+
+	// Create the parent key against which to seal the data
+	srkHandle, _, err := tpm2.CreatePrimary(rwc, tpm2.HandleOwner, tpm2.PCRSelection{}, "", srkPassword, srkTemplate)
+	if err != nil {
+		return nil, fmt.Errorf("can't create primary key: %v", err)
+	}
+	defer func() {
+		tpm2.FlushContext(rwc, srkHandle)
+	}()
+	glog.Infof("Created parent key with handle: 0x%x\n", srkHandle)
+
+	// Load the sealed data into the TPM.
+	objectHandle, _, err := tpm2.Load(rwc, srkHandle, srkPassword, publicArea, privateArea)
+	if err != nil {
+		return nil, fmt.Errorf("unable to load data: %v", err)
+	}
+	defer func() {
+		tpm2.FlushContext(rwc, objectHandle)
+	}()
+	glog.Infof("Loaded sealed data with handle: 0x%x\n", objectHandle)
+
+	// Create the authorization session
+	sessHandle, _, err := policyPCRPasswordSession(rwc, pcr, objectPassword)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get auth session: %v", err)
+	}
+	defer func() {
+		tpm2.FlushContext(rwc, sessHandle)
+	}()
+
+	// Unseal the data
+	unsealedData, err := tpm2.UnsealWithSession(rwc, sessHandle, objectHandle, objectPassword)
+	if err != nil {
+		return nil, fmt.Errorf("unable to Unseal data: %v", err)
+	}
+	return unsealedData, nil
+}
+
+// Returns session handle and policy digest.
+func policyPCRPasswordSession(rwc io.ReadWriteCloser, pcr int, password string) (tpmutil.Handle, []byte, error) {
+	// FYI, this is not a very secure session.
+	sessHandle, _, err := tpm2.StartAuthSession(
+		rwc,
+		tpm2.HandleNull,  /*tpmKey*/
+		tpm2.HandleNull,  /*bindKey*/
+		make([]byte, 16), /*nonceCaller*/
+		nil,              /*secret*/
+		tpm2.SessionPolicy,
+		tpm2.AlgNull,
+		tpm2.AlgSHA256)
+	if err != nil {
+		return tpm2.HandleNull, nil, fmt.Errorf("unable to start session: %v", err)
+	}
+	defer func() {
+		if sessHandle != tpm2.HandleNull {
```

Comment:
Hmm, when you call Unseal, won't this close the session before `tpm2.UnsealWithSession` call?
I'd leave flushing to the caller entirely.

Reply:
Done. 
It is interesting that we are returning a session handle only to allow the caller to flush it. It seems that the session is not actually used (in our context), and the only thing we care about is the policy.
In this case, it would make sense to change the signature and just return the policy and flush the handle in this function.

PTAL at the rest of the changes.

Reply:
I'm confused about how `tpm2.UnsealWithSession` works, with a flushed session handle.
Shouldn't it fail if session is no longer active?

I thought Unseal would put flushing of sessHandle in a defer, instead of flushing it before use?
Then returning the handle makes sense.

Reply:
You are right - code in this commit results in
+ ./tpmunseal --path-to-output cleartext.json
F0426 21:11:24.346156   20379 main.go:50] unable to Unseal data: warning code 0x18 : the 1st authorization session handle references a session that is not loaded
goroutine 1 [running]:

So, we need to defer flushing the session (next commit).